crates/dbsp/src/operator/communication/exchange.rs

```diff
@@ -31,7 +60,188 @@ use std::{
 // using the id as a key.  If there already is an `Exchange` with this id in
 // the store, created by another worker, a reference to that `Exchange` will
 // be used instead.
-circuit_cache_key!(local ExchangeId<T>(usize => Arc<Exchange<T>>));
+circuit_cache_key!(local ExchangeCacheId<T>(ExchangeId => Arc<Exchange<T>>));
+
+#[tarpc::service]
+trait ExchangeService {
+    async fn exchange(data: Vec<u8>, exchange_id: usize, sender: usize, receiver: usize);
+}
+
+type ExchangeId = usize;
+
+// Maps from an `exchange_id` to the `Inner` that implements the exchange.
+type ExchangeDirectory = Arc<RwLock<HashMap<ExchangeId, Arc<Inner>>>>;
+
+#[derive(Clone)]
+struct ExchangeServer(ExchangeDirectory);
+
+#[tarpc::server]
+impl ExchangeService for ExchangeServer {
+    async fn exchange(
+        self,
+        _: context::Context,
+        data: Vec<u8>,
+        exchange_id: ExchangeId,
+        sender: usize,
+        receiver: usize,
+    ) {
+        let inner = self.0.read().unwrap().get(&exchange_id).unwrap().clone();
```

Comment:
I suspect that the error handling strategy will have to be improved.
or perhaps the connection failures are sufficient for the coordinator to do something sensible?

Reply:
Error handling will have to be improved. This implementation does not yet attempt to be robust in the presence of failure.